sql/ttl/ttljob: TestRowLevelTTLJobRandomEntries failed

sql/ttl/ttljob.TestRowLevelTTLJobRandomEntries [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8768256?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8768256?buildTab=artifacts#/) on master @ [4f7756ff4f16f38c5780d5c8ad0f54768ba97573](https://github.com/cockroachdb/cockroach/commits/4f7756ff4f16f38c5780d5c8ad0f54768ba97573):


```
=== RUN   TestRowLevelTTLJobRandomEntries
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/f00a2dae801c4acc82659d1d8dd20fad/logTestRowLevelTTLJobRandomEntries4263550092
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestRowLevelTTLJobRandomEntries
    ttljob_test.go:884: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/f00a2dae801c4acc82659d1d8dd20fad/logTestRowLevelTTLJobRandomEntries4263550092
--- FAIL: TestRowLevelTTLJobRandomEntries (323.34s)
=== RUN   TestRowLevelTTLJobRandomEntries/random_1
    ttljob_test.go:788: test case: ttljob_test.testCase{desc:"random 1", createTable:"CREATE TABLE tbl (\n\tid UUID DEFAULT gen_random_uuid(),\n\trand_col_1 FLOAT8,\n\trand_col_2 OID,\n\ttext TEXT,\n\tPRIMARY KEY (id, rand_col_1, rand_col_2)\n) WITH (ttl_expire_after = '30 days', ttl_select_batch_size = 1, ttl_delete_batch_size = 34)", preSetup:[]string(nil), postSetup:[]string(nil), numExpiredRows:1921, numNonExpiredRows:41, numSplits:7, forceNonMultiTenant:false, expirationExpression:"", addRow:(func(*ttljob_test.rowLevelTTLTestJobTestHelper, *tree.CreateTable, time.Time))(nil)}
    ttljob_test.go:155: condition failed to evaluate within 45s: expected status succeeded, got running (error: )
    --- FAIL: TestRowLevelTTLJobRandomEntries/random_1 (63.25s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-sessions
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRowLevelTTLJobRandomEntries.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-24653

sql/ttl/ttljob.TestRowLevelTTLJobRandomEntries [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9521800?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9521800?buildTab=artifacts#/) on master @ [12548c5b62bc48645fc56e8e35c04294c630af43](https://github.com/cockroachdb/cockroach/commits/12548c5b62bc48645fc56e8e35c04294c630af43):


```
=== RUN   TestRowLevelTTLJobRandomEntries
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/f00a2dae801c4acc82659d1d8dd20fad/logTestRowLevelTTLJobRandomEntries721903496
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestRowLevelTTLJobRandomEntries
    ttljob_test.go:890: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/f00a2dae801c4acc82659d1d8dd20fad/logTestRowLevelTTLJobRandomEntries721903496
--- FAIL: TestRowLevelTTLJobRandomEntries (426.63s)
=== RUN   TestRowLevelTTLJobRandomEntries/random_5
    ttljob_test.go:794: test case: ttljob_test.testCase{desc:"random 5", createTable:"CREATE TABLE tbl (\n\tid UUID DEFAULT gen_random_uuid(),\n\trand_col_1 BYTES,\n\trand_col_2 BOX2D,\n\ttext TEXT,\n\tPRIMARY KEY (id, rand_col_1, rand_col_2)\n) WITH (ttl_expire_after = '30 days', ttl_select_batch_size = 1, ttl_delete_batch_size = 43)", preSetup:[]string(nil), postSetup:[]string(nil), numExpiredRows:1865, numNonExpiredRows:51, numSplits:2, forceNonMultiTenant:false, expirationExpression:"", addRow:(func(*ttljob_test.rowLevelTTLTestJobTestHelper, *tree.CreateTable, time.Time))(nil)}
    ttljob_test.go:163: condition failed to evaluate within 45s: expected status succeeded, got running (error: )
    --- FAIL: TestRowLevelTTLJobRandomEntries/random_5 (70.97s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #100079 sql/ttl/ttljob: TestRowLevelTTLJobRandomEntries failed [C-test-failure O-robot branch-release-23.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRowLevelTTLJobRandomEntries.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/ttl/ttljob.TestRowLevelTTLJobRandomEntries [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9623239?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9623239?buildTab=artifacts#/) on master @ [c08946af1f935d543a033de6b7b585c4d65288c4](https://github.com/cockroachdb/cockroach/commits/c08946af1f935d543a033de6b7b585c4d65288c4):


```
=== RUN   TestRowLevelTTLJobRandomEntries
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/f00a2dae801c4acc82659d1d8dd20fad/logTestRowLevelTTLJobRandomEntries828143312
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestRowLevelTTLJobRandomEntries
    ttljob_test.go:895: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/f00a2dae801c4acc82659d1d8dd20fad/logTestRowLevelTTLJobRandomEntries828143312
--- FAIL: TestRowLevelTTLJobRandomEntries (391.88s)
=== RUN   TestRowLevelTTLJobRandomEntries/random_3
    ttljob_test.go:799: test case: ttljob_test.testCase{desc:"random 3", createTable:"CREATE TABLE tbl (\n\tid UUID DEFAULT gen_random_uuid(),\n\trand_col_1 UUID,\n\trand_col_2 TIMETZ,\n\ttext TEXT,\n\tPRIMARY KEY (id, rand_col_1, rand_col_2)\n) WITH (ttl_expire_after = '30 days', ttl_select_batch_size = 1, ttl_delete_batch_size = 59)", preSetup:[]string(nil), postSetup:[]string(nil), numExpiredRows:1953, numNonExpiredRows:96, numSplits:3, forceNonMultiTenant:false, expirationExpression:"", addRow:(func(*ttljob_test.rowLevelTTLTestJobTestHelper, *tree.CreateTable, time.Time))(nil)}
    ttljob_test.go:169: condition failed to evaluate within 45s: expected status succeeded, got running (error: )
    --- FAIL: TestRowLevelTTLJobRandomEntries/random_3 (69.04s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #100079 sql/ttl/ttljob: TestRowLevelTTLJobRandomEntries failed [C-test-failure O-robot T-sql-sessions branch-release-23.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRowLevelTTLJobRandomEntries.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/ttl/ttljob.TestRowLevelTTLJobRandomEntries [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9767941?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9767941?buildTab=artifacts#/) on master @ [cfa9ede9a74bf2bc7f683b2a6e4adc12f150be35](https://github.com/cockroachdb/cockroach/commits/cfa9ede9a74bf2bc7f683b2a6e4adc12f150be35):


```
=== RUN   TestRowLevelTTLJobRandomEntries
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/f00a2dae801c4acc82659d1d8dd20fad/logTestRowLevelTTLJobRandomEntries3999241685
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestRowLevelTTLJobRandomEntries
    ttljob_test.go:895: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/f00a2dae801c4acc82659d1d8dd20fad/logTestRowLevelTTLJobRandomEntries3999241685
--- FAIL: TestRowLevelTTLJobRandomEntries (432.88s)
=== RUN   TestRowLevelTTLJobRandomEntries/random_2
    ttljob_test.go:799: test case: ttljob_test.testCase{desc:"random 2", createTable:"CREATE TABLE tbl (\n\tid UUID DEFAULT gen_random_uuid(),\n\trand_col_1 TIMESTAMPTZ,\n\trand_col_2 TIMETZ,\n\ttext TEXT,\n\tPRIMARY KEY (id, rand_col_1, rand_col_2)\n) WITH (ttl_expire_after = '30 days', ttl_select_batch_size = 1, ttl_delete_batch_size = 19)", preSetup:[]string(nil), postSetup:[]string(nil), numExpiredRows:1732, numNonExpiredRows:14, numSplits:7, forceNonMultiTenant:false, expirationExpression:"", addRow:(func(*ttljob_test.rowLevelTTLTestJobTestHelper, *tree.CreateTable, time.Time))(nil)}
    ttljob_test.go:169: condition failed to evaluate within 45s: expected status succeeded, got running (error: )
    --- FAIL: TestRowLevelTTLJobRandomEntries/random_2 (69.17s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #100079 sql/ttl/ttljob: TestRowLevelTTLJobRandomEntries failed [C-test-failure O-robot T-sql-sessions branch-release-23.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRowLevelTTLJobRandomEntries.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/ttl/ttljob/ttljob_test.go
+++ b/pkg/sql/ttl/ttljob/ttljob_test.go
@@ -166,25 +166,29 @@ func (h *rowLevelTTLTestJobTestHelper) waitForScheduledJob(
 		regex, err = regexp.Compile(expectedErrorRe)
 		require.NoError(t, err)
 	}
-	testutils.SucceedsSoon(t, func() error {
+	testutils.SucceedsWithin(t, func() error {
 		// Force newly created job to be adopted and verify it succeeds.
 		h.server.JobRegistry().(*jobs.Registry).TestingNudgeAdoptionQueue()
-		var status, errorStr string
-		if err := h.sqlDB.DB.QueryRowContext(
-			context.Background(),
-			query,
-		).Scan(&status, &errorStr); err != nil {
-			return errors.Wrapf(err, "expected to scan row for a job, got")
+		rows := h.sqlDB.QueryStr(t, query)
+		var actualStatuses []string
+		var actualErrors []string
+		for _, row := range rows {
+			actualStatus := row[0]
+			actualError := row[1]
+			if actualStatus == string(expectedStatus) && (regex == nil || regex.MatchString(actualError)) {
+				return nil
+			}
+			actualStatuses = append(actualStatuses, actualStatus)
+			actualErrors = append(actualErrors, actualError)
 		}
-
-		if status != string(expectedStatus) {
-			return errors.Newf("expected status %s, got %s (error: %s)", expectedStatus, status, errorStr)
-		}
-		if regex != nil && !regex.MatchString(errorStr) {
-			return errors.Newf("expected error matches %s, got %s", expectedErrorRe, errorStr)
-		}
-		return nil
-	})
+		return errors.Newf(`
+expectedStatus="%s"
+actualStatuses="%s"
+ expectedError="%s"
+  actualErrors="%s"`,
+			expectedStatus, strings.Join(actualStatuses, `", "`), expectedErrorRe, strings.Join(actualErrors, `", "`),
+		)
+	}, 3*time.Minute)
 }
 
 func (h *rowLevelTTLTestJobTestHelper) verifyNonExpiredRows(

